src/avram/callbacks.cr

```diff
@@ -1,4 +1,21 @@
 module Avram::Callbacks
+  macro included
+    def before_run
+    end
+
+    def after_run(_object)
+    end
+
+    def before_save
+    end
+
+    def after_save(_record : T) forall T
+    end
+
+    def after_commit(_record : T) forall T
```

Comment:
I see this module is included in `Avram::Operation` and I'm thinking that these callbacks don't make sense outside of database operations.  Could this module be refactored to allow declaring life cycle events that be subscribed to? Then, `Avram::Operation` could declare before and after run events and `Avram::SaveOperation` could declare the others. Something like:

```crystal
module Avram::Callbacks
  macro register_life_cycle_event(event_name, *args)
    macro {{ event_name.id }}(method_name)
      def {{ event_name.id }}(
        {% for param in args %}
          {{ param.id }},
        {% end %}
      )
        \{% if @type.methods.map(&.name).includes?(:before_run.id) %}
          previous_def
        \{% else %}
          super
        \{% end %}

        \{{ method_name.id }}
      end
    end
  end
end
```

And then in the operations:

```crystal
abstract class Avram::Operation
  include Avram::Callbacks

  register_life_cycle_event :before_run
  register_life_cycle_event :after_run, object
end

abstract class Avram::SaveOperation(T) < Avram::Operation
  register_life_cycle_event :before_save
  register_life_cycle_event :after_save, record : T
  register_life_cycle_event :after_commit, record : T
end
```

I have no idea if this actually works but this allows `Avram::Callbacks` to be a generic utility and not add methods to `Avram::Operation` that shouldn't be there.

Reply:
I made a repo with a somewhat working example of my thoughts https://github.com/matthewmcgarvey/crystal-callbacks

Reply:
I like the idea of subscribing to events. Right now there's quite a bit of duplication which I hope to cleanup before this PR is fully done. The one thing I'm unclear on is how we would differentiate between when a method would be called before certain things, or after. Like, this would not stop someone from writing `register_life_cycle_event :around_save`, but how would that work? I'll have to think on this more after I start to clean up some of the duplication. That might give me a little more clarity. Thanks for the suggestions!

Reply:
@jwoertink that repo linked above has been updated to work (kind of) as you are wanting. Base classes will call `register_event :run` and then there will be macro methods for `before_save`, `after_save`, and `around_save`.  The macros got quite hairy 🙈

Reply:
Sweet! I'll take a look. Now that I have all these specs passing, I need to start looking at cleaning up some of my duplication and adding in the new bits for this. It's about to get pretty crazy! 😂